File: app/src/main/java/com/lumiyaviewer/lumiya/render/picking/IntersectInfo.java
package com.lumiyaviewer.lumiya.render.picking;

import android.opengl.Matrix;
import com.lumiyaviewer.lumiya.slproto.types.LLVector4;

public class IntersectInfo {
    public final int faceID;
    public final boolean faceKnown;
    public final LLVector4 intersectPoint;
    public final float s;
    public final float t;
    public final float u;
    public final float v;

    public IntersectInfo(IntersectInfo intersectInfo, float[] floats2, int i) {
        this.intersectPoint = intersectInfo.intersectPoint;
        this.faceID = intersectInfo.faceID;
        this.s = intersectInfo.s;
        this.t = intersectInfo.t;
        this.faceKnown = intersectInfo.faceKnown;
        if (!this.faceKnown) {
            this.u = intersectInfo.u;
            this.v = intersectInfo.v;
            return;
        }
        float[] floats = new float[8];
        floats[0] = this.s;
        floats[1] = this.t;
        floats[3] = 1.0f;
        Matrix.multiplyMV(floats, 4, floats2, i, floats, 0);
        this.u = floats[4];
        this.v = floats[5];
    }

    public IntersectInfo(LLVector4 intersectPoint) {
        this.intersectPoint = intersectPoint;
        this.faceID = 0;
        this.u = 0.0f;
        this.v = 0.0f;
        this.s = 0.0f;
        this.t = 0.0f;
        this.faceKnown = false;
    }

    public IntersectInfo(LLVector4 intersectPoint, int faceID, float f, float f2) {
        this.intersectPoint = intersectPoint;
        this.faceID = faceID;
        this.u = f;
        this.v = f2;
        this.s = f;
        this.t = f2;
        this.faceKnown = true;
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/render/picking/IntersectInfo.kt
package com.lumiyaviewer.lumiya.render.picking

import android.opengl.Matrix
import com.lumiyaviewer.lumiya.slproto.types.LLVector4

class IntersectInfo {
    @JvmField val faceID: Int
    @JvmField val faceKnown: Boolean
    @JvmField val intersectPoint: LLVector4
    @JvmField val s: Float
    @JvmField val t: Float
    @JvmField val u: Float
    @JvmField val v: Float

    constructor(source: IntersectInfo, textureMatrix: FloatArray, offset: Int) {
        intersectPoint = source.intersectPoint
        faceID = source.faceID
        s = source.s
        t = source.t
        faceKnown = source.faceKnown
        if (faceKnown) {
            val coordinates = floatArrayOf(s, t, 0.0f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f)
            Matrix.multiplyMV(coordinates, 4, textureMatrix, offset, coordinates, 0)
            u = coordinates[4]
            v = coordinates[5]
        } else {
            u = source.u
            v = source.v
        }
    }

    constructor(intersectPoint: LLVector4) {
        this.intersectPoint = intersectPoint
        faceID = 0
        u = 0.0f
        v = 0.0f
        s = 0.0f
        t = 0.0f
        faceKnown = false
    }

    constructor(intersectPoint: LLVector4, faceID: Int, u: Float, v: Float) {
        this.intersectPoint = intersectPoint
        this.faceID = faceID
        this.u = u
        this.v = v
        s = u
        t = v
        faceKnown = true
    }
}
